README.md
# [Legacy] Scansreader 

**Warning**: This is legacy code, not maintained anymore, and depending on a discontinued [QIV](https://packages.debian.org/unstable/qiv) library by Adam Kopacz. It may however be ported to the [new QIV version](http://spiegl.de/qiv/)?



Scansreader is a "do the right thing" lean and mean unix/linux X C  program to read a set of images scanned from printed documents, such as  comics book, as confortably as possible. This page is at https://colas.nahaboo.net/Code/ScansReader

###   Goals 

 Scansreader was done to:  

-  **Be safe**, with no functionality to accidentaly modify or delete the images on disk.
-  **Be usable**, with only adapted functionalities for the task (no window mode, only fullscreen full width mode), with  streamlined interaction (scroll only vertically, without clicking the  mouse), automatic bookmarks in unlimited numbers...
-  **Be fast**, pure C, based on the fast [QIV](http://www.klografx.net/qiv/) general image viewer, with added prefetching of next image.

##   Philosophy 

 When reading a book, I do not want to be bothered having to get to the  mouse or touchpad, especially when potato-couching with my notebook, and have to aim for menus, scrollbars, buttons. Thus scansreader do not use any of this interaction method, and is always fullscreen.

It also do not provide things that belong to generic image manipulation  programs like the gimp: no thumbnails, no 2-page mode, no dialog box to  open files (you just run scansreader via your favorite desktop program).

Why the "S" in Scan*s*reader instead of the more english sounding  Scanreader? Well, there was no google hit with scansreader, but  scanreader was already the name of many products... so I went with the  less common name.

### Implementation 

 I could not find a program satisfying these criteria. All involved too  much interaction or option-setting to really read scans as painlessly as a real book.

I stumbled on a nice and fast linux image viewer [QIV](http://www.klografx.net/qiv/) by Adam Kopacz, and decided to use it as the base for scansreader,  removing a lot of its code and adding functionalities. It makes use of  the X, gdk and imlib libraries.

More details in the `README.txt` and `HISTORY.txt` files.

### License 

GPLv3

## Download & Installation 

Just compile it (`make`) and, install (`make install`) or copy the compiled `scansreader` executable in your path, e.g. in  `/usr/local/bin`.

##   User manual 

 Just run `scansreader`. it will read all the images in the  current directory, (or the directory or zip, cbz, rar or cbr archive  given as argument), and open the first one full screen, resized to the  full width of the screen.  

-  the **-l** option will re-open the last document that was opened the last time we quit the program.
-  **-L fr** will use the French language for online help texts.
-  **--help** prints the full usage, with more options

  Then, you can read your document with the following commands:  

-  **moving the mouse** will scroll the document, as **using the mousewheel**, or pressing the **up / down arrow** keys will only scroll vertically, and stop at the top and bottom of the image, indicating the end with a small yellow triangle at the bottom  right or top left of the image.
-  **SPACE** will "do the right thing": move down  the image, then go to the next one once at the bottom. If the image has  been zoomed in, and is wider than the screen, goes first to the right  then down to the left, etc...
-  **left-click** or **right arrow** key will go to next page,
-  **right-click** or **left arrow** key will go to previous page,
-  **q** or **Escape** will quit, appending the current position to the ~/.scansreader.log file
-  **p** and **n** will go to the previous or next saved place
-  **1,2,...,9,0** Goes to image at 0, 10, 20, ..., 90, 100% of current set.
-  **?** (or any other non-understood key) displays online help, with more commands



##   Other scans and comics readers 

Scansreader can be used to read digitized comics. You may want to look at similar programs: 

-  [Comic Reader](http://creader.sourceforge.net/) is similar to scansreader. In Python for linux.
-  [Powerchick](http://home.gna.org/powerchick/) another python reader
-  [Superchick](http://sacredchao.net/~piman/software.shtml) another python reader geared for manga
-  And see more of them on the [Wikipedia page](http://www.zcultfm.com/~comic/wiki/index.php/Comic_Viewers)



##   History 

- 2021-12-13 moved to GitHub fopr reference, as I remove the the mercurial site

- 2009-01-18, Wiki page + HG soureces on my public mercurial site

- v1.14

   2006-03-05, nothing changed in the code, only the packaging: 

  -  the .tgz includes now a directory
  -  compilation is now done on debian sarge (was woody)
  -  static executable is no longer included in distrib, but in        a separate tgz for a faster download of the distrib

- v1.13

   2005-02-16,  

  -  no more "random" option 
  -  TAB command to toggle between full width and full screen
  -  next image is prefetched and decoded while showing the current one for        speed
  -  Shift-Q now quits without saving in history

- v1.12

   2005-02-13,  

  -  can decode on the fly zip and rar archives

-  **v1.11** 2004-05-09, bugfix for prev/next

-  **v1.10** 2004-03-12, history file format change

-  **v1.9** 2003-12-07, p and n navigates in history

-  **v1.8** 2003-12-05, 

-  **v1.7** 2003-11-03, scrolls with mouse motions only, no need to click

-  **v1.6** 2003-11-02, french help, indicator when at top or bottom

-  **v1.5** 2003-11-01, 

-  **v1.4** 2003-10-29, 

-  **v1.3** 2003-07-02, removed windowed mode

-  **v1.2** 2003-07-01, first usable release



##   Testimonies 

"Astor" wrote (in French) on some forum after finding scansreader:

> Fran,cais: ![fr.png](https://colas.nahaboo.net/pub/System/FamFamFamFlagIcons/fr.png)
>
> scansreader est absolument génial et totalement parfait, c'est exactement ce qu'il me fallait !
>
> C'est léger, rapide, puissant, ergonomique, accessible, pratique, efficace, et en plus çà marche bien. Rien que çà. J'ai essayé sous Linux une tonne d'autres softs, mais aucun ne m'a jamais convenu, certains sont écrits en Java avec les pieds et mettent 20 secondes sur ma machine à afficher la première page d'un .cbz (scanreader est instantané), d'autres ont une interface envahissante et discutable (scansreader est plein écran, comme CDisplay, et pour moi c'est l'idéal), certains sont compliqués à installer (bon c'est bien le java mais livrer juste les .class et vas y pour la commande de lancement trouve tout seul les .jar en librairies qu'il faut, non merci).
>
> Bref, un must. Merci beaucoup à l'auteur.
>
> Merci aussi pour la version compilée en static, çà m'a bien aidé.

Here is my translation in English: ![us.png](https://colas.nahaboo.net/pub/System/FamFamFamFlagIcons/us.png)

> scansreader is absolutely awesome and totally perfect, it is exactly what I needed!
>
> It is light, fast, powerful, usable, accessible, handy, efficient, and  moreover it works well. Nothing less. I tried many other programs, but I was satisfied by none, some are  written in bad Java and take 20 seconds to show the first page of a .cbz (scansreader is instantaneous), other have a bulky and dubious  interface (scansreader, like CDisplay, is fullscreen, and for me it is  the ideal choice), some are complex to install (java is great, but  providing only the .class files and let the user figure the magic  command to launch to find the relevant .jar libraries, no thanks).
>
> In a nutshell, a must have. Thanks to the author.
>
> Thanks also for the statically compiled version, it helped me a lot.



To be honest, "Astor" later found a shortcoming of scansreader: it uses  the QIV image scaling functions that are fast but of low quality and can make noticeable artifacts (jagged lines) on some kind of pages.


HISTORY.txt
[1.14] 2006-03-05 Colas Nahaboo  <http://colas.nahaboo.net>
	* nothing changed in the code, only the packaging:
		- the .tgz includes now a directory
		- compilation is now done on debian sarge (was woody)
		- static executable is no longer included in distrib, but in
		  a separate tgz for a faster download of the distrib

[1.13] (b1.10) 2005-02-16  Colas Nahaboo  <http://colas.nahaboo.net>

	* no more "random" option 
	* TAB command to toggle between full width and full screen

	* image.c (scansreader_load_image): prefetches next image

2005-02-14  Colas Nahaboo  <http://colas.nahaboo.net>

	* event.c (scansreader_handle_event): Q now quits without saving pos

2005-02-13  Colas Nahaboo  <colas@aye.inria.fr>

	* options.c: lib/ contents moved in main dir

[1.12] 2005-02-13  Colas Nahaboo  <colas@aye.inria.fr> (build 1.12)

	* utils.c: argument can be a zipped (.cbz or .bdz) or rar-ed
	(.cbr or .bdr) archives of files, that will be uncompressed on
	the fly

	* build 1.8
	* main.c: ~/.scansreader.last is not used anymore, everything is done in
	the ~/.scansreader.log

[1.11] 2004-05-09  Colas Nahaboo  <colas@aye.inria.fr>

	* main.c: fix: previous/next position was broken in some cases

[1.10] 2004-03-12 Colas Nahaboo  <colas@nahaboo.net>

	* format of ~/.scansreader.log and ~/.scansreader.last files changed: one entry per
	  line, tab-separated fields

[1.9] 2003-12-07  Colas Nahaboo  <colas@aye.inria.fr>

	* main.c: 'p' and 'n' navigate in history of last positions
	  p ==> previous one, n ==> next one

2003-12-05  Colas Nahaboo  <colas@aye.inria.fr>

	* main.c: appends last positions to a ~/.scansreader.log

2003-11-03  Colas Nahaboo  <colas@aye.inria.fr>

	* event.c: scrolls vertically with mouse motions

2003-11-02  Colas Nahaboo  <colas@aye.inria.fr>

	* main.c (scansreader_handle_timer): the slide show mode now has a default
	of 10s, and scolls down before going to next page
	"," and "." makes this delay 1 second faster or slower

	* main.h (helpstrs_fr): aide francaise: option -L fr

	* image.c: help restructured. unknonw keys toggle help

	* event.c: if image is not wider than screen, left & right arrows
	go to next & previous images

	* image.c (scansreader_load_image): always raise window on new load
	when scolling, if try to move past top or bottom, show a little
	indicator.

	* main.c: saves last pos in ~/.scansreader.last on exit, as a file of 3
	lines: pwd / file path / page index
	Goes back to last save with -C on start

	* Makefile (STATUSBAR_FONT): bigger font

	* event.c (scansreader_handle_event): on some cygwin systems, an extra key
	press & release event is generated after another, with same times
	as the real release, so we ignore it

2003-11-01  Colas Nahaboo  <colas@aye.inria.fr>

	* image.c: display a "Loading ..." message on loding images

2003-10-29  Colas Nahaboo  <colas@aye.inria.fr>

	* event.c: better computation of scroll factor

2003-07-02  Colas Nahaboo  <colas@aye.inria.fr>

	* image.c (update_image): removed code for windowed mode

	* event.c (scansreader_handle_event): unknown key => online help

2003-07-01  Colas Nahaboo  <colas@aye.inria.fr>

	* event.c: move by keys moves twice slower
	(scansreader_handle_event): removed functions to set image as root

	* image.c (setup_win): cursor is always blank

	* Makefile: new ci target via fshar



main.c
/*
  Module       : main.c
  Purpose      : GDK/Imlib Quick Image Viewer (scansreader)
  More         : see scansreader README
  Homepage     : http://www.klografx.net/scansreader/
  Policy       : GNU GPL
*/	

#include <errno.h>
#include <gdk/gdkx.h>
#include <gtk/gtkwidget.h>
#include <stdio.h>
#include <signal.h>
#include <sys/time.h>
#include <ctype.h>
#include <string.h>
#include "scansreader.h"
#include "main.h"

scansreader_image main_img;

GdkImlibImage *next_im = NULL;			/* prefetched next image */
int next_image_idx = -1;

static void filter_images(int *, char **);
static int check_extension(const char *);
static void scansreader_signal_usr1();
static void scansreader_signal_usr2();
static gboolean scansreader_handle_timer(gpointer);
static void scansreader_timer_restart(gpointer);
void save_last_pos();
void do_the_right_thing (scansreader_image *);
static int initial_image_idx = 0;
extern int rreaddir(const char *);

int main(int argc, char **argv)
{
  struct timeval tv;

  //er, we don't set a value for params.dither here?
  //i would probably just let the user's .imrc set it
  //GdkImlibInitParams params;
  //params.flags = PARAMS_DITHER;

  /* Initialize GDK and Imlib */

  gdk_init(&argc,&argv);
  gdk_imlib_init();

  //probably should only use one of the init funcs...
  //gdk_imlib_init_params(&params);

  /* Load things from GDK/Imlib */

  scansreader_main_loop = g_main_new(TRUE);
  cmap = gdk_colormap_get_system();
  text_font = gdk_font_load(STATUSBAR_FONT);
  screen_x = gdk_screen_width();
  screen_y = gdk_screen_height();
  cursor = visible_cursor = gdk_cursor_new(CURSOR);

  gtk_widget_push_visual(gdk_imlib_get_visual());
  gtk_widget_push_colormap(gdk_imlib_get_colormap());

  /* Randomize seed for 'true' random */

  gettimeofday(&tv,NULL);
  srand(tv.tv_usec*1000000+tv.tv_sec);

  /* Set up our options, image list, etc */

  reset_mod(&main_img);
  helpstrs = helpstrs_us; helpkeys = helpstrs + 2;
  options_read(argc, argv, &main_img);
  max_rand_num = images;

  if (!strcmp(language, "fr")) {
    helpstrs = helpstrs_fr; helpkeys = helpstrs + 2;
  }

  if (filter) /* Filter graphic images */
    filter_images(&images,image_names);
  
  if (!images) { /* No images to display */
    g_print("scansreader: cannot load any images.\n");
    usage(argv[0],1);
  }

  /* get colors */

  color_alloc(STATUSBAR_BG, &text_bg);
  color_alloc(ERROR_BG, &error_bg);
  color_alloc(image_bg_spec, &image_bg);

  /* Display first image first, except in random mode */

  image_idx = 0;
  if (random_order) next_image(0);
  if (initial_image_idx) next_image(initial_image_idx);

  /* Setup callbacks */

  gdk_event_handler_set(scansreader_handle_event, &main_img, NULL);
  scansreader_timer_restart(NULL);

  /* And signal catchers */

  signal(SIGTERM, finish);
  signal(SIGINT, finish);
  signal(SIGUSR1, scansreader_signal_usr1);
  signal(SIGUSR2, scansreader_signal_usr2);

  /* Load & display the first image */

  scansreader_load_image(&main_img);
  g_main_run(scansreader_main_loop); /* will never return */
  return 0;
}


void scansreader_g_exit(int code, int save_position)
{
  if (save_position) save_last_pos();
  if (text_font) gdk_font_unref(text_font);
  if (cmap) gdk_colormap_unref(cmap);
  destroy_image(&main_img);

  g_main_destroy(scansreader_main_loop);
  finish(SIGTERM);		/* calls scansreader_exit(0) */
}

void scansreader_exit(int code)
{
  scansreader_tmpdir_clean();
  gdk_exit(code);
}


/*
 * functions for handling signals
 */

static void scansreader_signal_usr1()
{
  next_image(1);
  scansreader_load_image(&main_img);
}

static void scansreader_signal_usr2()
{
  next_image(-1);
  scansreader_load_image(&main_img);
}


/*
 * Slideshow timer function
 *
 * If this function returns false, the timer is destroyed
 * and scansreader_timer_restart() is automatically called, which
 * then starts the timer again. Thus images which takes some
 * time to load will still be displayed for "delay" seconds.
 */
 
static gboolean scansreader_handle_timer(gpointer data)
{
  if (*(int *)data || slide) {
    do_the_right_thing(&main_img);
      /*    next_image(0); */
  }
  return FALSE;
}


/*
 *	Slideshow timer (re)start function
 */
 
static void scansreader_timer_restart(gpointer dummy)
{
  g_timeout_add_full(G_PRIORITY_LOW, delay,
		     scansreader_handle_timer, &slide,
		     scansreader_timer_restart);
}

/* Filter images by extension */

static void filter_images(int *images, char **image_names)
{
  int i = 0;

  while(i < *images) {
    if (check_extension(image_names[i])) {
      i++;
    } else {
      int j = i;
      while(j < *images-1) {
	image_names[j] = image_names[j+1];
	++j;
      }
      --(*images);
    }
  }
}

static int check_extension(const char *name)
{
  char *extn = strrchr(name, '.');
  int i;

  if (extn)
    for (i=0; image_extensions[i]; i++)
      if (strcasecmp(extn, image_extensions[i]) == 0)
        return 1;

  return 0;
}

void remove_nl(char *s) {
  char *p = strrchr(s, '\n');if (p) *p = '\0';
}

void
save_last_pos() {
  FILE *fp;
  char savefile[1024];

  sprintf(savefile, "%s/.scansreader.log", getenv("HOME"));
  fp = fopen(savefile, "a");
  fprintf(fp, "%s\t%s\t%d\n", vchdir_pwd, image_names[image_idx], image_idx);
  fclose(fp);
}

static struct _place {
  int idx; char dir[1024]; char file[1024];
} *places;

void previous_place(int step) {
  static int cur_place;
  static int last_place;
  int new_place;
  
  if (places == NULL) {
    FILE *fp;
    char *homedir;
    char savefile[1024];
    char buffer[1024];
    int nmatch;
    homedir = getenv("HOME");
    sprintf(savefile, "%s/.scansreader.log", homedir);
    if ((fp = fopen(savefile, "r"))) {
      while (fgets(buffer, 1023, fp)) {
	places = realloc(places, (last_place + 1) * sizeof(struct _place));
	nmatch = sscanf(buffer, "%[^\t]\t%[^\t]\t%d\n", 
		  places[last_place].dir, places[last_place].file, 
			&(places[last_place].idx));
	if (nmatch != 3) {
	  g_print("scansreader: Warning: bad log line #%d: %s\n", last_place+1, 
		  buffer);
	  continue;
	}
	remove_nl(places[last_place].dir);
	remove_nl(places[last_place].file);
	cur_place = last_place;
	last_place++;
      }
    } else {
      cur_place = 0; last_place = 0;
    }
  }
  new_place = cur_place + step;
  if (new_place >= 0 && new_place < last_place) {
    cur_place = new_place;
    image_idx = initial_image_idx = places[new_place].idx;
    g_print("scansreader: resuming page %d in dir %s\n", 
	    initial_image_idx, places[new_place].dir);
    if (!vchdir(places[new_place].dir)) {
      if (image_names) {
	free(image_names);
	image_names = 0;
      }
      images = 0; max_image_cnt = 0;
      rreaddir(".");
    } else {
      g_print("scansreader: ***ERROR[%d]*** Could not go in dir: %s\n", 
	      errno, places[new_place].dir);
    }
  }
}

void
restore_last_pos() {
  previous_place(0);
}


main.h

#ifndef MAIN_H
#define MAIN_H

int		first = 1; /* TRUE iff this is first image */
char		infotext[BUF_LEN];
GdkCursor	*cursor, *visible_cursor, *invisible_cursor;
GMainLoop	*scansreader_main_loop;
gint		screen_x, screen_y; /* Size of the screen in pixels */
GdkFont		*text_font; /* font for statusbar and help text */
GdkColormap	*cmap; /* global colormap */
char		*image_bg_spec = IMAGE_BG;
GdkColor	image_bg; /* default background */
GdkColor	text_bg; /* statusbar and help backgrounf */
GdkColor	error_bg; /* for the error window/screen */
int		images;	/* Number of images in current collection */
char		**image_names; /* Filenames of the images */
int		image_idx; /* Index of current image displayed. 0 = 1st image */
int		max_image_cnt; /* # images currently allocated into arrays */
scansreader_deletedfile *deleted_files;
int		delete_idx;

/* Options and such */

int	filter = 1;
gint	center = 1;
gint	default_brightness = DEFAULT_BRIGHTNESS;
gint	default_contrast = DEFAULT_CONTRAST;
gint	default_gamma = DEFAULT_GAMMA;
gint	delay = SLIDE_DELAY; /* delay in slideshow mode in seconds */
int	random_order; /* TRUE if random delay in slideshow */
int	random_replace = 1; /* random with replacement by default */
int	recursive; /* TRUE if must recurse */
int	maxpect; /* TRUE if autozoom (fit-to-screen) mode */
int	statusbar_fullscreen = 1; /* TRUE if statusbar in fullscreen is turned on (default) */
int	statusbar_window = 0; /* FALSE if statusbar in window is turned off (default) */
int	slide; /* 1=slide show running */
int	scale_down; /* resize down if image x/y > screen */
int	transparency; /* transparency on/off */
int	do_grab; /* grab keboard/pointer (default off) */
int	max_rand_num; /* the largest random number range we will ask for */
int	fixed_window_size = 0; /* window width fixed size/off */
int	fixed_zoom_factor = 0; /* window fixed zoom factor (percentage)/off */
int zoom_factor = 0; /* zoom factor/off */

char *language = "us";
char **helpstrs;
/* For --help output, we'll skip the first two lines. */
char **helpkeys; /* helpstrs+2 */

/* Used for filtering */

const char *image_extensions[] = {
#ifdef EXTN_JPEG
    ".jpg",".jpeg",
#endif
#ifdef EXTN_GIF
    ".gif",
#endif
#ifdef EXTN_TIFF
    ".tif",".tiff",
#endif
#ifdef EXTN_XPM
    ".xpm",
#endif
#ifdef EXTN_XBM
    ".xbm",
#endif
#ifdef EXTN_PNG
    ".png",".pjpeg",
#endif
#ifdef EXTN_PPM
    ".ppm",
#endif
#ifdef EXTN_PNM
    ".pnm",
#endif
#ifdef EXTN_PGM
    ".pgm",
#endif
#ifdef EXTN_PCX
    ".pcx",
#endif
#ifdef EXTN_BMP
    ".bmp",
#endif
#ifdef EXTN_EIM
    ".eim",
#endif
#ifdef EXTN_TGA
    ".tga",
#endif
    NULL
};

#endif /* MAIN_H */


options.c
/*
  Module       : options.c
  Purpose      : Read and evaluate commandline options
  More         : see scansreader README
  Homepage     : http://www.klografx.net/scansreader/
*/	

#include "scansreader.h"
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>
#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#else
#include "getopt.h"
#endif

extern char *optarg;
extern int optind, opterr, optopt;
extern int rreaddir(const char *);
extern void restore_last_pos();
extern char *language;

static char *short_options = "hexyzmtb:c:g:niIpavo:srSd:u:fw:W:PMNlL:";
static struct option long_options[] =
{
    {"help",		0, NULL, 'h'},
    {"center",		0, NULL, 'e'},
    {"root",		0, NULL, 'x'},
    {"root_t",		0, NULL, 'y'},
    {"root_s",		0, NULL, 'z'},
    {"maxpect",		0, NULL, 'm'},
    {"scale_down",	0, NULL, 't'},
    {"brightness",	1, NULL, 'b'},
    {"contrast",	1, NULL, 'c'},
    {"gamma",		1, NULL, 'g'},
    {"language",        1, NULL, 'L'},
    {"no_filter",	0, NULL, 'n'},
    {"no_statusbar",	0, NULL, 'i'},
    {"statusbar",	0, NULL, 'I'},
    {"transparency",	0, NULL, 'p'},
    {"do_grab",		0, NULL, 'a'},
    {"version",		0, NULL, 'v'},
    {"bg_color",	1, NULL, 'o'},
    {"slide",		0, NULL, 's'},
    {"random",		0, NULL, 'r'},
    {"shuffle",         0, NULL, 'S'},
    {"delay",		1, NULL, 'd'},
    {"recursive",   	0, NULL, 'u'},
    {"fixed_width",		1, NULL, 'w'},
    {"fixed_zoom",		1, NULL, 'W'},
    {"ignore_path_sort",0, NULL, 'P'},
    {"last_pos",        0, NULL, 'l'},
    {0,			0, NULL, 0}
};

static int numeric_sort = 0, merged_case_sort = 1, ignore_path_sort = 0;

/* This array makes it easy to sort filenames into merged-case order
 * (e.g. AaBbCcDdEeFf...). */
static unsigned char casemap[256] = {
    0x00,0x01,0x02,0x03,0x04,0x05,0x06,0x07,
    0x08,0x09,0x0A,0x0B,0x0C,0x0D,0x0E,0x0F,
    0x10,0x11,0x12,0x13,0x14,0x15,0x16,0x17,
    0x18,0x19,0x1A,0x1B,0x1C,0x1D,0x1E,0x1F,
    0x20,0x21,0x22,0x23,0x24,0x25,0x26,0x27,
    0x28,0x29,0x2A,0x2B,0x2C,0x2D,0x2E,0x2F,
    0x30,0x31,0x32,0x33,0x34,0x35,0x36,0x37,
    0x38,0x39,0x3A,0x3B,0x3C,0x3D,0x3E,0x3F,
    0x40,0x41,0x43,0x45,0x47,0x49,0x4B,0x4D, /* @ABCDEFG */
    0x4F,0x51,0x53,0x55,0x57,0x59,0x5B,0x5D, /* HIJKLMNO */
    0x5F,0x61,0x63,0x65,0x67,0x69,0x6B,0x6D, /* PQRSTUVW */
    0x6F,0x71,0x73,0x75,0x76,0x77,0x78,0x79, /* XYZ[\]^_ */
    0x7A,0x42,0x44,0x46,0x48,0x4A,0x4C,0x4E, /* `abcdefg */
    0x50,0x52,0x54,0x56,0x58,0x5A,0x5C,0x5E, /* hijklmno */
    0x60,0x62,0x64,0x66,0x68,0x6A,0x6C,0x6E, /* pqrstuvw */
    0x70,0x72,0x74,0x7B,0x7C,0x7D,0x7E,0x7F, /* xyz{|}~  */
    0x80,0x81,0x82,0x83,0x84,0x85,0x86,0x87,
    0x88,0x89,0x8A,0x8B,0x8C,0x8D,0x8E,0x8F,
    0x90,0x91,0x92,0x93,0x94,0x95,0x96,0x97,
    0x98,0x99,0x9A,0x9B,0x9C,0x9D,0x9E,0x9F,
    0xA0,0xA1,0xA2,0xA3,0xA4,0xA5,0xA6,0xA7,
    0xA8,0xA9,0xAA,0xAB,0xAC,0xAD,0xAE,0xAF,
    0xB0,0xB1,0xB2,0xB3,0xB4,0xB5,0xB6,0xB7,
    0xB8,0xB9,0xBA,0xBB,0xBC,0xBD,0xBE,0xBF,
    0xC0,0xC1,0xC2,0xC3,0xC4,0xC5,0xC6,0xC7,
    0xC8,0xC9,0xCA,0xCB,0xCC,0xCD,0xCE,0xCF,
    0xD0,0xD1,0xD2,0xD3,0xD4,0xD5,0xD6,0xD7,
    0xD8,0xD9,0xDA,0xDB,0xDC,0xDD,0xDE,0xDF,
    0xE0,0xE1,0xE2,0xE3,0xE4,0xE5,0xE6,0xE7,
    0xE8,0xE9,0xEA,0xEB,0xEC,0xED,0xEE,0xEF,
    0xF0,0xF1,0xF2,0xF3,0xF4,0xF5,0xF6,0xF7,
    0xF8,0xF9,0xFA,0xFB,0xFC,0xFD,0xFE,0xFF
};

static int my_strcmp(const void *v1, const void *v2)
{
    unsigned char *cp1 = *(unsigned char **)v1;
    unsigned char *cp2 = *(unsigned char **)v2;
    unsigned char *sufptr1, *sufptr2;
    int tmp;

    sufptr1 = cp1 + strlen(cp1);
    while (--sufptr1 > cp1 && *sufptr1 != '.') {}
    sufptr2 = cp2 + strlen(cp2);
    while (--sufptr2 > cp2 && *sufptr2 != '.') {}

    if (ignore_path_sort) {
	unsigned char *slash;
	if ((slash = strrchr(cp1, '/')) != NULL)
	    cp1 = slash + 1;
	if ((slash = strrchr(cp2, '/')) != NULL)
	    cp2 = slash + 1;
    }
    if (numeric_sort) {
	int namelen = 0, diff = 0;

	do {
	    if (sufptr1 && (isdigit(*cp1) || isdigit(*cp2))) {
		unsigned char *ep1, *ep2;

		if (diff)
		    return diff;
		for (ep1 = cp1; isdigit(*ep1); ep1++) {}
		if (ep1 == cp1)
		    return 1;
		for (ep2 = cp2; isdigit(*ep2); ep2++) {}
		if (cp2 == ep2)
		    return -1;
		if ((diff = (ep1 - cp1) - (ep2 - cp2)) == 0) {
		    long val = atol(cp1) - atol(cp2);
		    diff = val < 0? -1 : val > 0? 1 : 0;
		}
		if (diff
		 && sufptr1 - cp1 <= namelen && sufptr2 - cp2 <= namelen)
		    return diff;
		namelen += ep1 - cp1;
		cp1 = ep1;
		cp2 = ep2 - 1;
	    }
	    else {
		if (cp1 == sufptr1) {
		    if (cp2 != sufptr2)
			return -1;
		    sufptr1 = sufptr2 = NULL;
		}
		else if (cp2 == sufptr2)
		    return 1;
		if (merged_case_sort)
		    tmp = casemap[*cp1++] - casemap[*cp2];
		else
		    tmp = *cp1++ - *cp2;
		if (tmp != 0)
		    return tmp;
		if (*cp2 == '/')
		    namelen = 0;
		else
		    namelen++;
	    }
	} while (*cp2++ != '\0');
	return diff;
    }

    do {
	if (cp1 == sufptr1) {
	    if (cp2 != sufptr2)
		return -1;
	}
	else if (cp2 == sufptr2)
	    return 1;
	if (merged_case_sort)
	    tmp = casemap[*cp1++] - casemap[*cp2];
	else
	    tmp = *cp1++ - *cp2;
	if (tmp != 0)
	    return tmp;
    } while (*cp2++ != '\0');

    return 0;
}

void options_read(int argc, char **argv, scansreader_image *q)
{
    int long_index, shuffle = 0, need_sort = 0;
    int c, cnt;
    int force_statusbar=-1;             /* default is don't force */
    int files_read = 0;

    /* BD mode */
    recursive = 0;
    q->win_w = fixed_window_size = screen_x;
    center=0;
    need_sort = 1;
    force_statusbar=0;

    while ((c = getopt_long(argc, argv, short_options,
	    long_options, &long_index)) != -1) {
	switch(c) {
	    case 'h': show_help(argv[0], 0); break;
	    case 'e': center=0;
		      break;
	    case 't': scale_down=1;
		      break;
	    case 'm': maxpect=1;
		      break;
	    case 'b': q->mod.brightness = (atoi(optarg)+32)*8;
		      if ((q->mod.brightness<0) || (q->mod.brightness>512))
			  usage(argv[0],1);
		      break;
	    case 'c': q->mod.contrast = (atoi(optarg)+32)*8;
		      if ((q->mod.contrast<0) || (q->mod.contrast>512))
			  usage(argv[0],1);
		      break;
	    case 'g': q->mod.gamma = (atoi(optarg)+32)*8;
		      if ((q->mod.gamma<0) || (q->mod.gamma>512))
			  usage(argv[0],1);
		      break;
	    case 'n': filter=0;
		      break;
	    case 'i': force_statusbar=0;
		      break;
	    case 'I': force_statusbar=1;
		      break;
	    case 'p': transparency=1;
		      break;
	    case 'a': do_grab=1;
		      break;
	    case 'v': g_print("ScansReader v%s\n", VERSION);
		      scansreader_exit(0);
		      break;
	    case 'o': image_bg_spec = optarg;
		      break;
	    case 's': slide=1;
		      break;
	    case 'r': random_order=1;
		      break;
	    case 'S': shuffle=1;
		      break;
	    case 'd': delay=atoi(optarg)*1000+1;
		      if (delay<1) usage(argv[0],1);
		      break;
	    case 'u': recursive = 1;
                      need_sort=1;
		      break;
	    case 'w': q->win_w = fixed_window_size = atoi(optarg);
		      break;
	    case 'W': fixed_zoom_factor = (atoi(optarg) - 100) / 10;
		      break;
	    case 'P': need_sort = ignore_path_sort = 1;
		      break;
   	    case 'l': restore_last_pos();
	      files_read = 1;
	      break;
  	    case 'L': language = optarg;
	      break;
	    case 0:
	    case '?': usage(argv[0], 1);
		      scansreader_exit(0);
	}
    }

    /* default: show statusbar only in fullscreen mode */
    /* user wants to override? */
    if (force_statusbar != -1) {
      statusbar_window=statusbar_fullscreen = force_statusbar;
    }

    /* if file argument, read all the files in its dir
     * if dir, read all its files
     * if no arg, read current dir
     */
    if (!files_read) {
      if((cnt = argc - optind) > 0) {
	char *name;
	struct stat sb;
	while (cnt-- > 0) {
	  name = argv[optind++];
	  if (stat(name, &sb) >= 0 ) {
	    if (!vchdir(name)) {		/* arg is dir */
	      rreaddir(".");
	    } else {			/* arg is file: go into its dir */
	      char dirname[1024]; char *filename;
	      strncpy(dirname, name, 1023); dirname[1023] = '\0';
	      filename = strrchr(dirname, '/');
	      if (filename) *filename = '\0';
	      else strcpy(dirname, ".");
	      if (!vchdir(dirname)) {
		rreaddir(".");
	      } else {
		g_print("scansreader: cannot go in dir: %s\n", dirname);
	      }
	    }
	  } else {
	    g_print("scansreader: cannot open argument: %s\n", name);
	  }
	}
      } else {
	vchdir_set_pwd();
	rreaddir(".");
      }
    }       
    if(shuffle) {
	char *fn;
	int i, p;
	for (i=0;i<images;i++) { /* simple insertion sort, fine for small num */
	  p = (int)(((float)rand()/RAND_MAX) * (images-i)) + i;
	  fn = image_names[i];
	  image_names[i] = image_names[p];
	  image_names[p] = fn;
	}
    }
    else if(need_sort)
	qsort(image_names, images, sizeof *image_names, my_strcmp);
}
